### PR TITLE
Hail2 rrm

### PR DESCRIPTION
**Pull request by catoverdrive**
@jbloom22 I put in a rough version of a Python `BlockMatrix.to_local_matrix()`. It won't do super large matrices---it can only accommodate `MAXINT / 8 - 3` entries---but I needed something like this for testing. I was looking at maybe removing `ComputeRRM` from Scala, but it looks like a lot of the lmmreg tests are using it, so I didn't touch it for now; do you know how easy that is work around?

### CODE REVIEW

File: python/hail/linalg/matrix.py [hunk 1]
```diff
@@ -108,5 +110,21 @@ def persist(self, storage_level='MEMORY_AND_DISK'):
     def unpersist(self):
         return BlockMatrix(self._jbm.unpersist())
 
-block_matrix_type.set(BlockMatrix)
+    @handle_py4j
+    def to_local_matrix(self):
```
Comment (jbloom22) [inline python/hail/linalg/matrix.py]: I think this should be to_numpy_matrix. I want to reserve to_local_matrix for converting a hail matrix from distributed to local.

File: python/hail/methods/tests.py [hunk 2]
```diff
@@ -241,6 +244,63 @@ def load_bin(ns, path):
...
+                                                m1,
+                                                fst=(k * [fst]),
+                                                seed=seed,
+                                                num_partitions=4,
+                                                af_dist=UniformDist(0.1, 0.9)).to_hail2()
+
+        def to_int_python(ds):
+            return BlockMatrix.from_matrix_table(ds['GT'].num_alt_alleles()).to_local_matrix()
+
+        def manual_calculation(ds):
```
Comment (jbloom22) [inline python/hail/methods/tests.py]: I have a hard time thinking of digital work as manual labor. :) How about `direct_calculation`?

File: python/hail/methods/statgen.py [hunk 3]
```diff
@@ -721,3 +722,106 @@ def grm(dataset):
...
+    ----------
+    call_expr : :class:`.CallExpression`
+        Expression on a :class:`.MatrixTable` that gives the genotype call.
+
+    Returns
+        :return: Realized Relationship Matrix for all samples.
+        :rtype: :class:`.KinshipMatrix`
+        """
+    source = call_expr._indices.source
+    if not isinstance(source, MatrixTable):
```
[the remark this answers is not in the document]
Comment (catoverdrive) [reply python/hail/methods/statgen.py]: I copied this from pca, but I'm not sure what it adds. Maybe a better error message? @tpoterba?

File: python/hail/methods/statgen.py [hunk 4]
```diff
@@ -721,3 +722,106 @@ def grm(dataset):
...
+        raise FatalError("Cannot run RRM: found 0 variants after filtering out monomorphic sites.")
+    info("Computing RRM using {} variants.".format(n_variants))
+
+    call_expr._indices = dataset._entry_indices
+    gt_expr = call_expr.num_alt_alleles()
+    normalized_genotype_expr = functions.bind(
+        dataset.AC / dataset.n_called,
+        lambda mean_gt: functions.bind(
+            functions.sqrt((dataset.ACsq +
+                            (n_samples - dataset.n_called) * mean_gt ** 2) /
```
[the remark this answers is not in the document]
Comment (catoverdrive) [reply python/hail/methods/statgen.py]: this looks fine to me? It looks like it's doing the python thing where it lines the second line of an expression up with the start of that expression?

File: src/main/scala/is/hail/utils/Py4jUtils.scala [hunk 5]
```diff
@@ -23,7 +26,18 @@ trait Py4jUtils {
       list.add(elem)
     list
   }
-  
+
+  // FIXME: Currently this can only send much less than the limit of
+  // BlockMatrix.toLocalMatrix. That's probably fine.
+  def doubleBDMToBytes(mat: BDM[Double]): Array[Byte] = {
+    // assumes matrix is written as written by BlockMatrix
```
Comment (jbloom22) [inline src/main/scala/is/hail/utils/Py4jUtils.scala]: BlockMatrix assumes offset = 0 and stride is minimal. But you do need to pass through the transpose state through as well, as done for read and write in RichDenseMatrixDouble, and then use that bit to decide whether to transpose on the numpy side.

File: src/main/scala/is/hail/utils/Py4jUtils.scala [hunk 6]
```diff
@@ -23,7 +26,18 @@ trait Py4jUtils {
...
+    val buf = new Array[Byte](8 + 8 * mat.rows * mat.cols)
+    Memory.storeInt(buf, 0, mat.rows)
+    Memory.storeInt(buf, 4, mat.cols)
+    Memory.memcpy(buf, 8, mat.data, 0, 8*mat.rows*mat.cols)
```
Comment (jbloom22) [inline src/main/scala/is/hail/utils/Py4jUtils.scala]: spaces around *

File: src/main/scala/is/hail/utils/Py4jUtils.scala (hunk 5 above)
Comment (jbloom22) [reply src/main/scala/is/hail/utils/Py4jUtils.scala]: add an fatal checking that `8 + 8 * mat.rows * mat.cols` doesn't overflow

File: python/hail/methods/tests.py [hunk 7]
```diff
@@ -7,6 +7,9 @@
 import numpy as np
 from struct import unpack
 import hail.utils as utils
+from hail.stats import UniformDist
```
Comment (jbloom22) [inline python/hail/methods/tests.py]: unused import

File: python/hail/linalg/matrix.py [hunk 8]
```diff
@@ -108,5 +110,21 @@ def persist(self, storage_level='MEMORY_AND_DISK'):
     def unpersist(self):
         return BlockMatrix(self._jbm.unpersist())
 
-block_matrix_type.set(BlockMatrix)
+    @handle_py4j
+    def to_numpy_matrix(self):
+        b = Env.jutils().doubleBDMToBytes(self._jbm.toLocalMatrix())
+        rows, cols = unpack('II', b[:8])
```
Comment (jbloom22) [inline python/hail/linalg/matrix.py]: you'll want to unpack the transpose state here as well, see below

File: src/main/scala/is/hail/utils/Py4jUtils.scala [hunk 9]
```diff
@@ -23,7 +26,18 @@ trait Py4jUtils {
       list.add(elem)
     list
   }
-  
+
+  // FIXME: Currently this can only send much less than the limit of
```
Comment (jbloom22) [inline src/main/scala/is/hail/utils/Py4jUtils.scala]: can you be more specific on the limit

File: src/main/scala/is/hail/utils/Py4jUtils.scala (hunk 5 above)
Comment (jbloom22) [reply src/main/scala/is/hail/utils/Py4jUtils.scala]: Also, please rename function `bdmToBytes`, as it's clear that it's doubles from the signature. And rename `mat` to the more specific `bdm` that's used elsewhere.
I'm also wary about not checking these assumptions. I suggest moving
`assertCompatibleLocalMatrix` out of BlockMatrix to the RichDenseMatrixDouble class, and changing the name to `assertCompatible()`?

Comment (catoverdrive) [reply src/main/scala/is/hail/utils/Py4jUtils.scala]: I'm putting these in as asserts and moving the actual error handling to the python side.

I think we could also make this handle the offset != 0 and stride != col/row case pretty easily, but that might not be necessary for our use case. What do you think?

Comment (jbloom22) [reply src/main/scala/is/hail/utils/Py4jUtils.scala]: I think we shouldn't add that complexity unless/until we need it.

File: python/hail/methods/tests.py [hunk 10]
```diff
@@ -261,6 +260,8 @@ def test_rrm(self):
 
         def direct_calculation(ds):
             ds = BlockMatrix.from_matrix_table(ds['GT'].num_alt_alleles()).to_numpy_matrix()
+            print(ds.shape)
```
Comment (jbloom22) [inline python/hail/methods/tests.py]: remove prints

File: src/main/scala/is/hail/utils/Py4jUtils.scala [hunk 11]
```diff
@@ -27,14 +27,13 @@ trait Py4jUtils {
...
-  // BlockMatrix.toLocalMatrix. That's probably fine.
-  def doubleBDMToBytes(mat: BDM[Double]): Array[Byte] = {
-    // assumes matrix is written as written by BlockMatrix
-    val buf = new Array[Byte](8 + 8 * mat.rows * mat.cols)
-    Memory.storeInt(buf, 0, mat.rows)
-    Memory.storeInt(buf, 4, mat.cols)
-    Memory.memcpy(buf, 8, mat.data, 0, 8*mat.rows*mat.cols)
+  // FIXME: Currently this needs 8 * r * c < int.max.
+  def bdmToBytes(bdm: BDM[Double]): Array[Byte] = {
+    assert(8 * bdm.rows * bdm.cols < Integer.MAX_VALUE)
```
Comment (jbloom22) [inline src/main/scala/is/hail/utils/Py4jUtils.scala]: change 8 to 8L or else this won't work

File: src/main/scala/is/hail/utils/Py4jUtils.scala [hunk 12]
```diff
@@ -27,14 +27,13 @@ trait Py4jUtils {
...
 
-  // FIXME: Currently this can only send much less than the limit of
-  // BlockMatrix.toLocalMatrix. That's probably fine.
-  def doubleBDMToBytes(mat: BDM[Double]): Array[Byte] = {
-    // assumes matrix is written as written by BlockMatrix
-    val buf = new Array[Byte](8 + 8 * mat.rows * mat.cols)
-    Memory.storeInt(buf, 0, mat.rows)
-    Memory.storeInt(buf, 4, mat.cols)
-    Memory.memcpy(buf, 8, mat.data, 0, 8*mat.rows*mat.cols)
+  // FIXME: Currently this needs 8 * r * c < int.max.
```
Comment (jbloom22) [inline src/main/scala/is/hail/utils/Py4jUtils.scala]: The factor of 8 is definitely going to rear it's head, it's the difference between 5792 and 46340 for side of square matrix.

Now that you have the assert, I think the FIXME should be made more specific about what to do. Send the data in smaller chunks, yeah?

File: python/hail/utils/java.py [hunk 13]
```diff
@@ -127,11 +127,19 @@ def numpy_from_breeze(bdm):
...
         raise ValueError("Expected major stride of Breeze matrix to be {}, found {}"
                          .format(expected_stride, bdm.majorStride()))
-    if cols * rows > 0x7ffff:
+    if cols * rows > 0x7ffffff:
         raise ValueError("rows * cols must be smaller than {}, found {} by {} matrix"
-                         .format(0x7ffff, rows, cols))
+                         .format(0x7ffffff, rows, cols))
 
-    b = Env.jutils().bdmToBytes(bdm)
+    if cols * rows <= 0x7ffff:
```
Comment (jbloom22) [inline python/hail/utils/java.py]: why did you choose 0x7ffff?

File: src/main/scala/is/hail/utils/Py4jUtils.scala [hunk 14]
```diff
@@ -26,14 +26,13 @@ trait Py4jUtils {
       list.add(elem)
     list
   }
-
-  // FIXME: Currently this needs 8 * r * c < int.max.
-  def bdmToBytes(bdm: BDM[Double]): Array[Byte] = {
-    assert(8 * bdm.rows * bdm.cols < Integer.MAX_VALUE)
+  
+  def bdmGetBytes(bdm: BDM[Double], start: Int, n: Int): Array[Byte] = {
+    assert(8L * n < Integer.MAX_VALUE.toLong)
```
Comment (jbloom22) [inline src/main/scala/is/hail/utils/Py4jUtils.scala]: .toLong not needed on right

File: python/hail/utils/java.py [hunk 15]
```diff
@@ -127,11 +127,19 @@ def numpy_from_breeze(bdm):
     if bdm.majorStride() != expected_stride:
         raise ValueError("Expected major stride of Breeze matrix to be {}, found {}"
                          .format(expected_stride, bdm.majorStride()))
-    if cols * rows > 0x7ffff:
+    if cols * rows > 0x7ffffff:
```
Comment (jbloom22) [inline python/hail/utils/java.py]: max int has seven Fs

Comment (jbloom22) [reply python/hail/utils/java.py]: i'd add `entries = rows * cols` and use that throughout

File: python/hail/utils/java.py (hunk 13 above)
Comment (catoverdrive) [reply python/hail/utils/java.py]: Maxint/8

Comment (jbloom22) [reply python/hail/utils/java.py]: that has floor 0xFFFFFFF

Comment (catoverdrive) [reply python/hail/utils/java.py]: oops, you're right. I shifted by 3 x 4 bits instead of by 3 bits.

**Comment by catoverdrive** [comment]
@jbloom I passed a 500x500 matrix and a 5000x5000 matrix of zeros from Scala. The 500x500 matrix took ~.3 seconds, and the 5000x5000 matrix took ~30 seconds to pass through. Almost all of this is spent retrieving the byte array from Java -> Python, so I don't know if we'll be able to get this noticeably faster.

The 0x100000 threshold for breaking up the blocks of bytes (semi-arbitrarily) because larger blocks caused me to run out of heap space, but wiggling that value up and down seems not to really affect the amount of time it takes.

**Pull request merged**